### PR TITLE
PoS Validator Constants and Network Changes

### PR DESCRIPTION
**Pull request by AeonSw4n**

### CODE REVIEW

**Comment by AeonSw4n** [comment]
Current dependencies on/for this PR:
* `feature/pos-networking-and-syncing`
  * **PR #876** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/876?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #768** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/768?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #857** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/857?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #858** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/858?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #859** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/859?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #860** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/860?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #861** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/861?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #792** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/792?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #793** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/793?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/876?utm_source=stack-comment).

File: lib/server.go [hunk 1]
```diff
@@ -1144,6 +1146,8 @@ func (srv *Server) _handleSnapshot(pp *Peer, msg *MsgDeSoSnapshotData) {
 		"<%v>, Last entry: <%v>), (number of entries: %v), metadata (%v), and isEmpty (%v), from Peer %v",
 		msg.SnapshotChunk[0].Key, msg.SnapshotChunk[len(msg.SnapshotChunk)-1].Key, len(msg.SnapshotChunk),
 		msg.SnapshotMetadata, msg.SnapshotChunk[0].IsEmpty(), pp)))
+	// Free up a slot in the operationQueueSemaphore, now that a chunk has been processed.
+	srv.snapshot.FreeOperationQueueSemaphore()
```
Comment (lazynina) [inline lib/server.go:1150]: Is this the correct place to do this? Should it happen towards the end of the function? Would it make sense to defer this?

File: lib/snapshot.go [hunk 2]
```diff
@@ -471,8 +471,6 @@ func (snap *Snapshot) Run() {
 				operation.blockHeight); err != nil {
 				glog.Errorf("Snapshot.Run: Problem adding snapshot chunk to the db")
 			}
-			// Free up a slot in the operationQueueSemaphore, now that a chunk has been processed.
-			<-snap.operationQueueSemaphore
```
Comment (lazynina) [inline lib/snapshot.go:475]: I'm not the expert on freeing up the slots here, so I'm not sure about this stuff.

**Review by lazynina — CHANGES_REQUESTED** [review]
Overall looks pretty good - just a few minor nits.

File: lib/constants.go [hunk 3]
```diff
@@ -478,6 +478,22 @@ func GetEncoderMigrationHeightsList(forkHeights *ForkHeights) (
 	return migrationHeightsList
 }
 
+type ProtocolVersionType uint64
+
+const (
+	ProtocolVersion0 ProtocolVersionType = 0
+	ProtocolVersion1
+	ProtocolVersion2
```
Comment (tholonious) [inline lib/constants.go:486]: Make these explicit so it's clear to future readers
```suggestion
	ProtocolVersion1 ProtocolVersionType = 1
	ProtocolVersion2 ProtocolVersionType = 2
```

**Review by tholonious — CHANGES_REQUESTED** [review]

File: lib/server.go (hunk 1 above)
Comment (AeonSw4n) [reply lib/server.go:1150]: Yeah, it's fine to free the chunk queue as soon as we receive a more-or-less valid chunk. The logic below this point checks for the chunk's validity; if we find it invalid, we generally disconnect the peer or restart the node. If the chunk is valid, we process it, which is expensive. Freeing the channel here allows us to get the next chunk a little bit faster compared to if we postponed this operation to the end of the function.

File: lib/snapshot.go (hunk 2 above)
Comment (AeonSw4n) [reply lib/snapshot.go:475]: The existing approach was prone to deadlocking, which caused integration tests to fail. I'm using the integration tests later in the stack, so I had to address the deadlock.

**Review by tholonious — APPROVED** [review]

**Pull request merged**